pydicom/data/download.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2018-2020 pydicom authors. See LICENSE file for details.
+
+import functools
+import json
+import os
+import pathlib
+import urllib.request
+import warnings
+import hashlib
+import contextlib
+
+try:
+    import tqdm
+
+    class DownloadProgressBar(tqdm.tqdm):
+        def update_to(self, b=1, bsize=1, tsize=None):
+            if tsize is not None:
+                self.total = tsize
+            self.update(b * bsize - self.n)
+
+except ImportError:
+    @contextlib.contextmanager
+    def DownloadProgressBar(*args, **kwargs):
+        try:
+            class dummy:
+                def update_to(*args, **kwargs):
+                    pass
+
+            yield dummy
+        finally:
+            pass
+
+HERE = pathlib.Path(__file__).resolve().parent
+
+
+def calculate_file_hash(filename):
+    BLOCKSIZE = 65536
+    hasher = hashlib.sha256()
+    with open(filename, "rb") as f:
+        buf = f.read(BLOCKSIZE)
+        while len(buf) > 0:
+            hasher.update(buf)
+            buf = f.read(BLOCKSIZE)
+
+    return hasher.hexdigest()
+
+
+def get_config_dir():
+    config_dir = pathlib.Path.home().joinpath(".pydicom")
+    config_dir.mkdir(exist_ok=True)
+
+    return config_dir
+
+
+def download_with_progress(url, filepath):
+    with DownloadProgressBar(
+        unit="B", unit_scale=True, miniters=1, desc=url.split("/")[-1]
+    ) as t:
+        urllib.request.urlretrieve(url, filepath, reporthook=t.update_to)
+
+
+def get_data_dir():
+    data_dir = get_config_dir().joinpath("data")
+    data_dir.mkdir(exist_ok=True)
+
+    return data_dir
+
+
+@functools.lru_cache()
+def get_url_map():
+    with open(HERE.joinpath("urls.json"), "r") as f:
+        url_map = json.load(f)
+
+    return url_map
+
+
+def get_url(filename):
+    url_map = get_url_map()
+
+    try:
+        url = url_map[filename]
+    except KeyError:
+        raise ValueError(
+            "The file provided isn't within pydicom's urls.json record.")
+
+    return url
+
+
+def download_all():
+    paths = []
+    for file_name in get_url_map().keys():
+        paths.append(data_path(file_name))
+
+    return paths
+
+
+def data_path(
+        filename,
```

Comment:
I think this could use a better name, to indicate that it will download the file... I don't know, maybe `data_path_with_download`?  Clumsy, but at least gives the indication that it is happening.

Reply:
All good :), I like it. Done.